[RELAY][OP] Operators.  	 pool2d, global_pool2d, batch_flatten, tanh, sigmoid, floor, ceil, trunc, abs, negative, multiply, mod, pow,  resize

srkreddy1238 opened the pull request:
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

junrushao1994 commented on python/tvm/relay/op/transform.py:
```diff
@@ -0,0 +1,46 @@
+from __future__ import absolute_import as _abs
+from . import _make
+
+def flatten(data):
```
The `flatten` here is not consistent with numpy's flatten (https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.flatten.html). Why not use names like `batch_flatten`?

tqchen replied:
+1 for batch_flatten

junrushao1994 commented on src/relay/op/tensor/transform.cc:
```diff
@@ -0,0 +1,77 @@
+  if (data == nullptr) return false;
+  if (data->shape.size() == 0) return false;
+
+  auto target_dim = make_const(Int(32), 1);
```
Will it be better to use `Int(64)`?

srkreddy1238 replied:
Done.

srkreddy1238 replied:
Default type is Int(32) for all indexing till now.
Using Int(64) require explicit cast on the result shape (TensorType) comparison.

junrushao1994 replied:
@srkreddy1238 I see. Thanks!

tqchen commented on include/tvm/relay/attrs/nn.h:
```diff
@@ -67,6 +67,101 @@ struct ConvAttrs : public tvm::AttrsNode<ConvAttrs> {
+  Array<IndexExpr> padding;
+  std::string layout;
+  IndexExpr ceil_mode;
+  IndexExpr count_include_pad;
```
we can use bool, if this does not have to be None

tqchen commented on include/tvm/relay/attrs/nn.h:
```diff
@@ -67,6 +67,101 @@ struct ConvAttrs : public tvm::AttrsNode<ConvAttrs> {
+};
+
+/*! \brief Attributes for upsampling operator */
+struct UpSamplingAttrs : public tvm::AttrsNode<UpSamplingAttrs> {
```
This belongs to vision.h

tqchen commented on include/tvm/relay/attrs/nn.h:
```diff
@@ -67,6 +67,101 @@ struct ConvAttrs : public tvm::AttrsNode<ConvAttrs> {
+
+/*! \brief Attributes for upsampling operator */
+struct UpSamplingAttrs : public tvm::AttrsNode<UpSamplingAttrs> {
+  IndexExpr scale;
```
can be float, if it does not have to be None

tqchen commented on python/tvm/relay/__init__.py:
```diff
@@ -10,6 +10,7 @@
 from .op import Op
 from .op.tensor import *
 from .op import nn
+from .op import transform
```
We only need namespace nn, vision. Normal numpy API should be in the root namespace

tqchen commented on python/tvm/relay/op/tensor.py:
```diff
@@ -60,6 +61,125 @@ def sqrt(data):
+        The computed result.
+    """
+    return _make.trunc(data)
+
```
two lines between global functions

tqchen commented on python/tvm/relay/op/tensor.py:
```diff
@@ -94,7 +214,76 @@ def subtract(lhs, rhs):
+    return _make.subtract(lhs, rhs)
+
+def multiply(lhs, rhs):
+    """Elementwise multiplication.
```
This is not elementwise operation, all operations broadcast

tqchen commented on python/tvm/relay/op/transform.py:
```diff
@@ -0,0 +1,46 @@
+
+    return _make.flatten(data)
+
+def resize(data,
```
this belongs to image.py

tqchen commented on python/tvm/relay/op/transform.py:
```diff
@@ -0,0 +1,46 @@
+           layout="NCHW",
+           method="BILINEAR",
+           align_corners=False):
+    """Resize operator.
```
document the behavior as it is less clear here

tqchen commented on tests/python/relay/test_op_level2.py:
```diff
@@ -58,5 +57,134 @@ def test_conv2d_infer_type():
+    x = ib.param("x", relay.ty.TensorType((n, c, h, w), "int8"))
+    th, tw = tvm.var("th"), tvm.var("tw")
+
+    with ib.function(x) as func:
```
create test_op_levelk(where k equals the corresponding level, and put the files onto their respective positions.

tqchen commented on src/relay/op/tensor/elemwise.cc:
```diff
@@ -106,6 +182,63 @@ RELAY_REGISTER_OP("subtract")
+    return CallNode::make(op, {lhs, rhs}, Attrs(), {});
+  });
+
+RELAY_REGISTER_OP("divide")
```
since BINARY_BCAST operator is soo common, create a macro for it just like unary operations.

tqchen commented on src/relay/op/tensor/elemwise.cc:
```diff
@@ -106,6 +182,63 @@ RELAY_REGISTER_OP("subtract")
+  .set_num_inputs(2)
+  .add_argument("lhs", "Tensor", "The left hand side tensor.")
+  .add_argument("rhs", "Tensor", "The right hand side tensor.")
+  .set_support_level(1)
```
be mindful about the support level, not everything is one

tqchen commented on src/relay/op/nn/pooling.cc:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  std::vector<IndexExpr> oshape({dshape[0], dshape[1], dshape[2], dshape[3]});
+  oshape[hidx] = tvm::select((param->ceil_mode == 1),
```
can directly use bool in the ceil mode

tqchen commented on include/tvm/relay/attrs/vision.h:
```diff
@@ -14,7 +14,7 @@ namespace relay {
 /*! \brief Attributes for upsampling operator */
 struct UpSamplingAttrs : public tvm::AttrsNode<UpSamplingAttrs> {
-  double scale;
+  int scale;
```
we should still keep this as double, as there can be fractional scale size, but try to make sure the inference deal wth it

srkreddy1238 replied:
Ref. https://github.com/dmlc/tvm/blob/0247b97198b0fecf82e6ca44d2f34bc8dfc3540f/nnvm/include/nnvm/top/nn.h#L378
It's an int in NNVM. Please confirm to use double here.

The test case here gets tricky with ```int32((float32(100)*2.000000f))``` in side shape !!

tqchen commented:
Please keep adding more docs to the python side of docstring as a followup of #1821.
Please also note some namespace convention change.  The operators that can have better docs include:

- All nn operators and upsample, try to document the computation rule
- batch_flatten, try to document its behavior by writing down a computation rule and convention.

tqchen commented on docs/langref/relay_op.rst:
```diff
@@ -35,14 +42,32 @@ This level enables typical convnet models.
+   tvm.relay.nn.avg_pool2d
+   tvm.relay.nn.global_max_pool2d
+   tvm.relay.nn.global_avg_pool2d
+   tvm.relay.upsampling
```
upsampling should belong to relay.image? as it is image operator

srkreddy1238 replied:
In NNVM it's under nn.
cc @masahi

masahi replied:
why not tvm.relay.nn.upsampling?

upsampling is a standard nn op.

srkreddy1238 replied:
ok @masahi

tqchen commented on docs/langref/relay_op.rst:
```diff
@@ -51,8 +76,28 @@ Level 1 Definitions
+
+Level 5 Definitions
+-------------------
+.. autofunction:: tvm.relay.vision.resize
```
do we need to distinguish between image and vision? as resize is more a image operator

srkreddy1238 replied:
Tricky call as we have very few ops in each category.

masahi replied:
In topi, resize belongs to topi.image.

srkreddy1238 replied:
Can I take conclusion to go with new category image for resize?

tqchen commented on docs/langref/relay_op.rst:
```diff
@@ -51,8 +76,28 @@ Level 1 Definitions
+
+Level 3 Definitions
+-------------------
+.. autofunction:: tvm.relay.batch_flatten
```
batch_flatten should belong to level2 as it is necessary for convnet

tqchen commented on src/relay/op/tensor/broadcast.cc:
```diff
@@ -0,0 +1,69 @@
+.add_type_rel("Broadcast", BroadcastRel);
+
+// Power
+RELAY_REGISTER_BINARY_BCAST_OP("pow")
```
power support level is 4

tqchen commented:
@siju-samuel can you also help review this?

siju-samuel commented on include/tvm/relay/attrs/nn.h:
```diff
@@ -67,6 +67,80 @@ struct ConvAttrs : public tvm::AttrsNode<ConvAttrs> {
+
+  TVM_DECLARE_ATTRS(AvgPool2DAttrs, "relay.attrs.AvgPool2DAttrs") {
+    TVM_ATTR_FIELD(pool_size)
+      .describe("Size of the pooling windows..");
```
windows.. -> windows.

siju-samuel commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,116 @@ def conv2d(data,
+
+
+def upsampling(data,
+               scale=1,
```
scale=1 default value needed?

srkreddy1238 replied:
fallback value ...

siju-samuel commented on python/tvm/relay/op/tensor.py:
```diff
@@ -83,7 +211,80 @@ def add(lhs, rhs):
+
 
 
 def subtract(lhs, rhs):
```
So many blank lines

siju-samuel commented on python/tvm/relay/op/vision/image.py:
```diff
@@ -0,0 +1,41 @@
+    ----------
+    data : relay.Expr
+        The input data to the operator.
+
```
parameter size is missing in docstring

siju-samuel commented on src/relay/op/nn/pooling.cc:
```diff
@@ -0,0 +1,272 @@
+
+  if (data == nullptr) return false;
+  auto dshape = data->shape;
+  if (dshape.size() ==  0) return false;
```
instead of 'if' check and 'return false', suggest to use 'CHECK' here, if there is no issue.. 
will be helpful for developers to locate issue.

siju-samuel commented on src/relay/op/nn/pooling.cc:
```diff
@@ -0,0 +1,272 @@
+    return false;
+  }
+
+  std::vector<IndexExpr> oshape({dshape[0], dshape[1], dshape[2], dshape[3]});
```
just use oshape = dshape;

srkreddy1238 replied:
I think we don't have constructor for this kind of assignment.

siju-samuel commented on src/relay/op/nn/pooling.cc:
```diff
@@ -0,0 +1,272 @@
+
+RELAY_REGISTER_OP("nn.avg_pool2d")
+.describe(R"code(
+Average pooling operation for one dimensional data.
```
one dimensional?

srkreddy1238 commented:
- Hope addressed all the review comments.

- TODO: Computation rule documentation pending. I would send as another PR considering multiple people working on the relay ops and may result in conflicts & rebase.

srkreddy1238 commented:
cc @siju-samuel

srkreddy1238 commented:
Docs updated.

srkreddy1238 commented:
cc @junrushao1994 @siju-samuel @tqchen

tqchen commented on docs/langref/relay_op.rst:
```diff
@@ -35,24 +41,71 @@ This level enables typical convnet models.
+   tvm.relay.nn.global_max_pool2d
+   tvm.relay.nn.global_avg_pool2d
+   tvm.relay.nn.upsampling
+   tvm.relay.batch_flatten
```
batch_flatten should be an nn operator as it is only used by nn

tqchen commented:
please rebase again to the most recent master

tqchen commented:
@srkreddy1238 please rebase against the latest master

tqchen commented on include/tvm/relay/attrs/image.h:
```diff
@@ -0,0 +1,41 @@
+  Array<IndexExpr> size;
+  std::string layout;
+  std::string method;
+  IndexExpr align_corners;
```
if it is concrete and do not have None value, use bool instead

srkreddy1238 replied:
Done.

srkreddy1238 commented:
@tqchen  ready for  review.

tqchen commented on docs/langref/relay_op.rst:
```diff
@@ -89,4 +134,10 @@ Level 4 Definitions
+Level 5 Definitions
+-------------------
+.. autofunction:: tvm.relay.image.resize
+>>>>>>> [RELAY][OP] Operators.
```
remove this line

tqchen commented on python/tvm/relay/op/image/image.py:
```diff
@@ -0,0 +1,42 @@
+           layout="NCHW",
+           method="BILINEAR",
+           align_corners=False):
+    """Resize.
```
Image resize operator

tqchen commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,215 @@ def conv2d(data,
+               padding=(0, 0),
+               layout="NCHW",
+               ceil_mode=False):
+    r"""MaxPool2D.
```
2D maximum pooling operator

tqchen commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,215 @@ def conv2d(data,
+        Layout of the input.
+
+    ceil_mode : bool, optional
+        To enable or disable ceil while pooling.
```
Returns
----------

tqchen commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,215 @@ def conv2d(data,
+
+    count_include_pad : bool, optional
+        To include padding to compute the average.
+    """
```
Returns
----------

tqchen commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,215 @@ def conv2d(data,
+
+    layout : str, optional
+        Layout of the input.
+
```
Returns
----------

tqchen commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,215 @@ def conv2d(data,
+
+    method : str, optional
+        Scale method to used [NEAREST_NEIGHBOR, BILINEAR].
+    """
```
Returns
----------

tqchen commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,215 @@ def conv2d(data,
+    return _make.upsampling(data, scale, layout, method)
+
+def batch_flatten(data):
+    """BatchFlatten.
```
flatten all the dimensions except for the batch dimension

tqchen commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,215 @@ def conv2d(data,
+
+    Returns
+    -------
+    tensor: The Flattened  tensor.
```
result: relay.Expr
     The flattened result

tqchen commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -86,3 +86,215 @@ def conv2d(data,
+    -------
+    tensor: The Flattened  tensor.
+    """
+
```
remove newline

tqchen commented on src/relay/op/nn/pooling.cc:
```diff
@@ -0,0 +1,270 @@
+
+TVM_REGISTER_NODE_TYPE(MaxPool2DAttrs);
+
+template <typename T>
```
T-> AttrTtype

tqchen commented on src/relay/op/nn/pooling.cc:
```diff
@@ -0,0 +1,270 @@
+  const auto* data = types[0].as<TensorTypeNode>();
+
+  CHECK(data != nullptr);
+  auto dshape = data->shape;
```
Use const auto& when possible

tqchen commented:
please also rebase against the master

srkreddy1238 commented:
cc @tqchen review pls.

tqchen commented:
Thanks @srkreddy1238 @siju-samuel @junrushao1994 @masahi this is now merged

srkreddy1238 commented:
Thank you all for reviews :)
